PRCS/prc-2.md

```diff
@@ -0,0 +1,174 @@
+It MUST initialize a lock similarly as in the `lock` function, and it MUST emit the `Lock` event
+
+```solidity
+interface IERC721Receiver {
+    /// @dev Whenever an {IERC721} `tokenId` token is transferred to this contract via {IERC721-safeTransferFrom}
+    ///  by `operator` from `from`, this function is called.
+    ///  It must return its Solidity selector to confirm the token transfer.
+    ///  If any other value is returned or the interface is not implemented by the recipient, the transfer will be
+    ///  reverted.
+    ///  The selector can be obtained in Solidity with `IERC721Receiver.onERC721Received.selector`.
+    ///  Note: the contract address is always the message sender.
+    /// @param _operator The address which called `safeTransferFrom` function
+    /// @param _from The address which previously owned the token
+    /// @param _tokenId The NFT identifier which is being transferred
+    /// @param _data Additional data with no specified format
+    /// @return `bytes4(keccak256("onERC721Received(address,address,uint256,bytes)"))`
+    ///  unless throwing
+    function onERC721Received(address _operator, address _from, uint256 _tokenId, bytes _data) external returns(bytes4);
+}
+```
+
+## Rationale
+
+The Hololocker contract is in its core a NFT staking contract. While there have been attempts at EIPs on this matter, such as [EIP-4987](https://eips.ethereum.org/EIPS/eip-4987), none gained enough traction and approval to be finalized.
+This proposal therefore proposes a standard for light-weight implementation of such mechanism.
+
+The main functions accept an array of token addresses and token IDs to be able to accomodate manipulations with multiple NFTs in one transaction, to save on transaction costs.
+
+The standard user interaction flow is: Lock -> Request Unlock -> (after `lockTime` has passed from the unlock request) Withdraw
+
+### Why Unlocking phase / lockTime ?
+
+If we did not have a special "Unlocking" state that requires users to wait for finality on the L1 before withdrawing their NFT, it could cause a situation where the same NFT is used in two places at once.
+Therefore, there must be a period of Unlocking state, initiated by the request for unlock and lasting the amount of seconds specified in the `lockTime` variable. This `lockTime` delay should reflect the finality period of the chain. Only after this delay has passed, the NFT can be withdrawn.
+
+### Consistent contract address
+
+To facilitate a more reliable experience for users, the Hololocker contract should be deployed to the same address across all EVM chains. This can be done by using a deployment proxy. Our reference implementation does this by specifiying a salt `bytes32(uint256(1))` in the [Foundry deployment script](https://github.com/dcSpark/projected-nft-whirlpool/blob/8b0d0367139eb9a43be94edff34a656258e25793/evm/script/Deploy.s.sol).  

```

Comment:
Do we want to use '1' for the salt? I don't think it matters I guess

Reply:
For us it doesn't matter what salt we use. Yes, with changed contract code comes a different address.